### ISSUE TITLE
Slow import of mzML files

### ISSUE THREAD

**Issue by tomas-pluskal** [mzmine/mzmine2#148]
Hi,

I often see the mzML import tasks stuck at 0% for VERY long time. This happens especially when multiple files are being imported from a network-mounted location (e.g., Windows SMB share). The code seems to be stuck in the indexing section of the jmzml library, but I was not able to track it to a particular problem.

First, I would like to know if anyone else has seen this problem. @dyrlund ? @golgauth ?

**Comment by dyrlund** [comment]
Isn't this related to [MSDK issue # 27](https://github.com/msdk/msdk/issues/27)?

**Comment by tomas-pluskal** [comment]
Possibly related, yes. But I'm seeing delays of over 30 minutes when importing relatively small (300 MB) files. I think that's unacceptable..

**Comment by dyrlund** [comment]
I agree that 30 minutes is not acceptable. This could be due to slow network transfer speed. Is yours 1Gbps or 100Mbps or wifi? 300MB is not much but if you e.g. process 16 files at the same time then you will have to transfer ~ 5GB of data which can take up to 30 minutes if the connection is though wifi and if others are also accessing data from the same network drive.

I have a similar speed issue on one of our workstations where the detection of the file type from a network storage can take 2-3 seconds pr. file even though we are running 1Gbps network. It can therefore take up to 30 minutes to add 500 mzML files to the task queue. On an identical computer this process takes less than a minute. I checked the transfer speed and they can both transfer files from the server with about 120MB/s.

**Comment by tomas-pluskal** [comment]
I am currently connected over 100Mbit ethernet. In fact after a bit of testing, I found it takes over 2 hours to open a single 300 MB mzML file over the network. The same file can be copied over the network in a few seconds, and parsed from the local drive in another few seconds.

I tracked the issue down to the xxindex dependency of jmzml (https://code.google.com/p/pride-toolsuite/wiki/XXIndex). This project provides the indexing function for mzML. Unfortunately, it does so in quite a stupid way: by reading single bytes.  See here:

https://github.com/PRIDE-Utilities/xxindex/blob/master/src/main/java/psidev/psi/tools/xxindex/index/XmlXpathIndexer.java#L161

This is extremely slow, at least in my case (network SMB share mounted to Mac OS X). The performance might differ depending on platform or network speed, but still I think it is unacceptable. 

I am considering to completely dump jmzml and use our own mzML parser... :(

**Comment by tomas-pluskal** [comment]
Nils @nilshoffmann , 
Have you ever encountered this problem?

**Comment by nilshoffmann** [comment]
Well, I never tried reading mzML over the network, but I did notice that the time between opening the file and being able to access it with jmzml is quite significant, even for rather small files. It think that it is possible to pass an existing xxindex file to the library to speed things up, and jmzml may also profit from indexexmzml, if the file is created with that option from proteowizard. For Maltcms, I ended up converting the mzML data to a custom netCDF dialect to be able to efficiently due random access on mass spectra and other information in the files. The startup time may be acceptable for batch processing of files, but it is really problematic for interactive applications.
